tensorflow/core/common_runtime/gpu/gpu_device.cc

```diff
@@ -1382,7 +1382,12 @@ Status BaseGPUDeviceFactory::CreateDevices(
                          std::make_pair(priority_low, priority_high)));
 #endif
     }
-    // Reset to the original device.
+    // Reset to the original device, if it is a valid visible gpu. Otherwise use
+    // the first visible device.
+    if (std::find(visible_gpu_order.begin(), visible_gpu_order.end(),
+                  original_device) == visible_gpu_order.end()) {
+      original_device = visible_gpu_order[0].value();
```

Comment:
I think it's possible for `visible_gpu_order` to be empty, e.g. if the user specifies an empty visible device list, in which case this accesses the array out of bounds. I think this means `cudaSetDevice` was not called previously, so you should skip calling it again in this case.

Also, using the first element of `valid_platform_device_ids` is better than the first element of `visible_gpu_order`. The former is equivalent to the latter except the former has filtered out some GPUs, e.g. GPUs whose compute capability is too low.

Reply:
Switched to use valid_platform_device_ids[0] as fallback. There is no concern about out of bounds access, since we are within the block.
```
if (!valid_platform_device_ids.empty()) {
    ...
}
```

Reply:
You're right, I missed that block.